src/main/java/com/beust/jcommander/JCommander.java

```diff
@@ -469,7 +469,7 @@ private List<String> expandDynamicArg(String arg) {
         for (ParameterDescription pd : descriptions.values()) {
             if (pd.isDynamicParameter()) {
                 for (String name : pd.getParameter().names()) {
-                    if (arg.startsWith(name) && !arg.equals(name)) {
+                    if (arg.startsWith(name) && !arg.equals(name) && arg.contains(pd.getParameter().getAssignment())) {
```

Comment:
This change removes an existing feature: It is valid to write `-abc` for a parameter `-ab` with a value of `c`. The actual bug is not that we forgot to check for the assignment operator, but that we must always use the *longest* argument name. In your example, `-abc` is longer than `-ab`, so the resolution must be `-abc` and not `-ab c`.

Reply:
```
@Test
  public void validateTest() {
    class Arg {
      @Parameter(names = { "-ab" })
      private String ab;
    }
    Arg arg = new Arg();
    JCommander jc = new JCommander(arg);
    jc.parse("-abc");
    Assert.assertEquals(arg.ab, "c");
  }
```
`com.beust.jcommander.ParameterException: Was passed main parameter '-abc' but no main parameter was defined in your arg class`

That's in master and without my change, if you can point towards a test that would fail that would help or create if possible. Currently all tests pass.

Else you need to provide me more details regarding the existing feature so that I can create a correct test case
> existing feature: It is valid to write -abc for a parameter -ab with a value of c

I tried with DynamicParamater `-a` and that `b` = `c`

Reply:
I've added a test way back but it might not cover what you mentioning.
```
public class ShadowingParameter {

    @Parameter(
            names = "-abc"
    )
    String abc;
    @DynamicParameter(
            names = "-ab"
    )
    Map<String,String> ab = new HashMap<>();

}
```

```
@Test
  public void shouldNotBreakSimilarParameter() {
    ShadowingParameter shadowingParameter = new ShadowingParameter();
    new JCommander(shadowingParameter).parse("-abc", "abc", "-ab", "a=b");
    Assert.assertEquals(shadowingParameter.abc, "abc");
    Assert.assertEquals(shadowingParameter.ab.get("a"), "b");
  }
```

This validate that `-abc` is treated as `-abc` and that `-ab`, `a=b` is treated as `-ab`
Without the change it would treat `-abc` as `-ab` and complain about the `c`

Reply:
I am very sorry for the confusion. It was solely my personal fault. I did not clearly enough read what your original bug description was like. Okay, understood, your fix does *not* break an existing feature then and I will be happy to merge it! :-)